services/venice-server/src/main/java/com/linkedin/venice/server/VeniceServer.java

```diff
@@ -714,7 +714,13 @@ private Function<String, Boolean> functionToCheckWhetherStorageEngineShouldBeKep
     return storageEngineName -> {
       String storeName = Version.parseStoreFromKafkaTopicName(storageEngineName);
 
-      AbstractStorageEngine storageEngine = storageService.getStorageEngine(storageEngineName);
```

Comment:
`checkWhetherStorageEngineShouldBeKeptOrNot` determines whether or not the entire `storageEngine` should be dropped, and where it's called it assumes that the `storageEngine` hasn't been opened yet.

What we should do here is modify the `StorageService` constructor to take another function that determines whether a partition should be kept or not. Ex: `checkWhetherStoragePartitionShouldBeKeptOrNot`

This new function can be invoked [here](https://github.com/linkedin/venice/blob/04aeaebb1fa56b5f392336074120ac9eb72acff1/clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/StorageService.java#L256C13-L256C55) after the `storageEngine` has been opened.

Reply:
This idea makes sense. I will integrate this.

Reply:
Operations involving Helix [`HelixParticipationService`, `SafeHelixDataAccessor`] may require that services are started for `StorageService.` [requires that Venice server has started]. `SafeHelixDataAccessor` is used to obtain the ideal state as referenced in #650. Will have to check the structure here so that `checkWhetherStoragePartitionsShouldBeKeptOrNot` executes correctly.

Reply:
Try moving the execution of `checkWhetherStoragePartitionsShouldBeKeptOrNot` to after `HelixParticipationService` is initialized in VeniceServer.

Reply:
(Resolution is here: https://github.com/linkedin/venice/pull/1196#discussion_r1781747306; when `HelixParticipationService` is started, `managerFuture` is completed in `HelixParticipationService.`).